2996_reflection/reflection.md

```diff
@@ -2283,6 +2284,86 @@ Lastly, we clarify that during the definition of an _injected declaration_, the
+
+```cpp
+template <std::meta::info R> consteval bool tfn() {
+  define_aggregate(R, {});
+  return true;
+}
+
+struct S;
+constexpr bool b = tfn<^^S>();
+  // OK, both manifestly constant-evaluated expression tfn<^^S>() and target scope of
+  // injected declaration for 'S' are in the global namespace
+```
+
+Nor does this rule prevent a class template from producing a declaration whose target scope is the same specialization.
+
+```cpp
+template <typename> struct TCls1 {
+  struct Incomplete;
+
+  consteval {
+    define_aggregate(^^Incomplete, {});
+      // OK, Incomplete is in the same instantiation as the define_aggregate call
+  }
+
+  static constexpr bool b = false;
+};
+
+template <typename T> struct TCls2 {
+  static consteval bool sfn()  // #1
+      requires (TCls1<T>::b) {
+    return false;  // never selected
+  }
+
+  static consteval bool sfn()  // #2
+      requires (true) {
+    return true;   // always selected
+  }
+};
+
+static_assert(TCls<void>::sfn());
+```
+
+Athough the instantiation of `TCls1<void>` in the requires-clause of `#1` causes an injected declaration to be produced, it is not discernibly a side-effect of the failed substitution: Observing the side effect will first require one to write (some moral  equivalent of) `TCLs1<void>::Incomplete`, the act of which would otherwise itself trigger the same side-effect.
+
+Although this rule constrains the manner with which `define_aggregate` can be used, we are not aware of any motivating use cases for P2996 that are harmed. Worth mentioning, however: the rule has more dire implications for other code injection papers being considered by WG21, most notably P3294 ("_Code Injection With Token Sequences_"). With this rule as it is, it becomes impossible for e.g., the instantiation of a class template specialization `TCls<Foo>` to produce an injected declaration of `std::formatter<TCls<Foo>>` (since the target scope would be the global namespace).
+
+In this context, we do believe that relaxations of the rule can be considered: For instance, we ought to be able to say that the instantiation of `std::formatter<TCls<Foo>>` is sequenced strictly after the instantiation of `TCls<Foo>`, and observations such as these might make it possible to permit such injections without making it "discernible" whether they resulted from failed substitutions. The key to such an approach would be to define a partial order over the instantiations of a program, and to extend the _semantically sequenced_ relation introduced by P2996 ([lex.phases]/7) to apply to constructs _across_ instantiations when the relative order of their respective instantiations is defined.
```

Comment:
say "this paper" or "this proposal" instead of P2996.

Reply:
Done